sim/include/proto/userValue.h

```diff
@@ -0,0 +1,179 @@
+            CurrentType;
+        // It matches one of the types
+        if (proto::isValOfType<CurrentType>(val)) {
+            return true;
+        }
+
+        // It doesn't match any type, recurse and get the next type
+        return _isValOfTypes<N + 1, AllowedTypes...>(val);
+    }
+};
+}  // namespace
+
+template <class... AllowedTypes>
+requires(proto::ProtobufType<AllowedTypes>&&...) bool isValOfTypes(
+    const bento::protos::Value& val) {
+    return _isValOfTypes<0, AllowedTypes...>(val);
+}
+
+namespace {
+template <size_t N = 0, class Type, class... TypeList>
+constexpr bool _typeInTypes() {
+    if constexpr (N == sizeof...(TypeList)) {
+        return false;
+    } else {
+        typedef std::tuple_element_t<N, std::tuple<TypeList...>> CurrentType;
+        if constexpr (std::is_same_v<Type, CurrentType>) {
+            return true;
+        }
+
+        return _typeInTypes<N + 1, Type, TypeList...>();
+    }
+}
+}  // namespace
+
+template <class Type, class... TypeList>
+constexpr bool typeInTypes() {
+    return _typeInTypes<0, Type, TypeList...>();
+}
+
+template <class... AllowedTypes, class Fn>
+bento::protos::Value runFnWithVal(const bento::protos::Value& x, Fn fn) {
+    if (!isValOfTypes<AllowedTypes...>(x)) {
+        throw std::runtime_error("Value given is not valid for this function.");
+    }
+
+    auto val = bento::protos::Value();
+    bool valSet = false;
+    if constexpr (typeInTypes<INT32, AllowedTypes...>()) {
+        if (isValOfType<INT32>(x)) {
+            setVal(val, fn(getVal<INT32>(x)));
+            valSet = true;
+        }
+    }
+    if constexpr (typeInTypes<INT64, AllowedTypes...>()) {
+        if (isValOfType<INT64>(x)) {
+            setVal(val, fn(getVal<INT64>(x)));
+            valSet = true;
+        }
+    }
+    if constexpr (typeInTypes<FLOAT32, AllowedTypes...>()) {
+        if (isValOfType<FLOAT32>(x)) {
+            setVal(val, fn(getVal<FLOAT32>(x)));
+            valSet = true;
+        }
+    }
+    if constexpr (typeInTypes<FLOAT64, AllowedTypes...>()) {
+        if (isValOfType<FLOAT64>(x)) {
+            setVal(val, fn(getVal<FLOAT64>(x)));
+            valSet = true;
+        }
+    }
+    if constexpr (typeInTypes<STR, AllowedTypes...>()) {
+        if (isValOfType<STR>(x)) {
+            setVal(val, fn(getVal<STR>(x)));
+            valSet = true;
+        }
+    }
+    if constexpr (typeInTypes<BOOL, AllowedTypes...>()) {
+        if (isValOfType<BOOL>(x)) {
+            setVal(val, fn(getVal<BOOL>(x)));
+            valSet = true;
+        }
+    }
+
+    if (!valSet) {
+        throw std::runtime_error("No value when executing function.");
+    }
+
+    return val;
+}
+
+template <class... AllowedTypes, class Fn>
+bento::protos::Value runFnWithVal(const bento::protos::Value& x,
+                                  const bento::protos::Value& y, Fn fn) {
+    auto val = bento::protos::Value();
+    runFnWithVal<AllowedTypes...>(x, [&val, &y, fn]<class X>(X x) {
```

Comment:
This function needs even more comments. Why are there two wrapping lambdas? Why are there 2 recursive calls?

Reply:
The first lambda is used to deduce the template type of the first parameter, `x`. The second lambda deduces the second template type of `y`. Now that we have deduced both types, we can call `fn`.